libs/galley-types/src/Galley/Types/Teams.hs

```diff
@@ -149,13 +149,25 @@ rolePermissions :: Role -> Permissions
 rolePermissions role = Permissions p p where p = rolePerms role
 
 permissionsRole :: Permissions -> Maybe Role
-permissionsRole (Permissions p p') | p /= p' = Nothing
+permissionsRole (Permissions p p')
+  | p /= p' =
+    -- we never did use @p /= p'@ for anything, fingers crossed that it doesn't occur anywhere
+    -- in the wild.  but if it does, this implementation prevents privilege escalation.
+    let p'' = Set.intersection p p'
+     in permissionsRole (Permissions p'' p'')
 permissionsRole (Permissions p _) = permsRole p
   where
     permsRole :: Set Perm -> Maybe Role
     permsRole perms =
       Maybe.listToMaybe
-        [role | role <- [minBound ..], rolePerms role == perms]
+        [ role
+          | (perms', role) <- reverse . sortBy (compare `on` (length . fst)) $ (\r -> (rolePerms r, r)) <$> [minBound ..],
```

Comment:
I'm a bit confused why we're sorting on the length of the permission set?  does that really do what we want?  We can have two permission sets that are the same size; but we must consider one permission set strictly less permissive than another (Perhaps based on the ordering of `Perm` ?)


e.g. the `rolePerms` for `RoleAdmin` and `RoleMember` have the same length; they both contain 4 elements; but `RoleMember <= RoleAdmin`

Reply:
Shouldn't we sort by the natural ordering of `Role` ? e.g. just iterate over `[minBound ..]`

So:


```
[ role |  role <- [minBound..],  rolePerms role `Set.isSubsetOf` perms]
```

Reply:
good idea, in this particular case it's the same, and the tests will show if it changes: c67cd00da2bd52128c458f88a0fc1a91d5bd0dcf